### DISCUSSION TITLE
Can my `loader` detect the original HTTP request method?

### DISCUSSION THREAD

**Discussion by aaronadamsCA** [remix-run/remix#6972] [Q&A]
I want my `loader` to behave differently depending on whether the **original** client request method was `GET` or `POST`. Does anyone know if this is possible?

When you `POST` to a route that uses `useActionData`, the `loader` needs to run as well. Remix seems to use a POST-redirect-GET pattern internally for this.

The problem is that I don't want this route to accept `GET` requests. I want to catch those and redirect them to another route. But I can't do it in the `loader`, because this is triggered on both `POST` and `GET` requests, and the loader's `request.method` is always `GET`.

My current workaround is to issue that fallback redirect on the client side as well, but I wonder if anyone has a better idea? For example, is there a way to make Remix directly render a route's `POST` instead of redirecting to `GET`, so that the loader will see `request.method` vary?

**Comment by kiliman** [comment]
Hmm.. Remix will call your `loader` for GET/HEAD requests. It will call your `action` for everything else POST/PUT/DELETE/etc. If a `loader` or `action` is not present for those requests, it will return `405 Method Not Allowed`.

The `useActionData()` hook only returns data that is returned from the `action` call. It never initiates a request. Same with `useLoaderData()`.

Remix will revalidate your routes after an `action` call. However, it will not revalidate a route that doesn't have a `loader`. Since you have a `loader` to handle the redirect case, you need to tell Remix it shouldn't revalidate this route. Simply export a `shouldRevalidate()` function that returns false.

https://remix.run/docs/en/1.19.1/route/should-revalidate

**Reply by aaronadamsCA** [reply]
Interesting. It sounds to me like the behaviour I'm seeing is probably unknown/undocumented, then, because what you describe is more or less what I see - except that Remix seems to internally redirect the browser's `POST` request to a `GET` request first.

I'm guessing it might work this way because otherwise there's no apparent way to progressively enhance routing with fetch; if a route component needs its own `loader` data to render, that is going to be a `GET` request to the route, as a part of serving the response to the `POST` request.

I am leaning toward this being a feature request of some sort - for instance, it seems to me that a `redirect()` call in a leaf route `loader` should probably be ignored while serving a non-`GET` request. Or maybe the `loader` should receive the true `method` of the client request, as opposed to always receiving `GET`.

I'll try to put together a reproduction when I have the time. It's particularly painful when issuing client-side redirects in embedded Shopify apps, which I believe is now a priority use case for Remix.
